### PR TITLE
Fix loss of precision on small angles in qua's pow

### PR DESCRIPTION
**Pull request by qsantos**
I encountered loss of precision when using quaternions with very small angles with the function `pow()`. This is because it determines the angle in the quaternion by applying `acos()` to the scale component `w`, which is equal to `1.` because of dramatic cancellation.

However, it is still possible to recover an accurate angle by looking at the non-scale components (which are then very close to, but different than, `0.`) and using `asin()`. I have written a fix to verify that this was possible, and it might be useful to merge it into the project.

I have tried to take non-normalized quaternions into account, but I may have missed something. Any feedback is welcome.

**Edit:** I have included a similar fix for qua's `angle()` function

### CODE REVIEW

**Comment by Groovounet** [comment]
The PR is failing the unit tests. Please, merge master to your branch and resolve the failure for the PR to be merged.

Thanks,
Christophe

**Comment by qsantos** [comment]
So, I have investigated this a bit further.

I had gone a bit fast with the threshold value. I have worked a bit more on this, and wrote a small program to iterate over all possible angles in single-precision floats from 0 to π / 2 to compare the loss of precision of the cos-method and the sin-method. It seems like the sin-method is worth it for angles roughly blow `1 rad`, and the cos-method for angles above this value.

This means the value `w` should be compared to `cos(0.5) ≈ 0.878`, not to `0.5`. The unit test which was failing uses an angle of π/2, or a quaternion with `w ≈ 0.707`, and expects an exact value (the result, which lies between `1.0` and `2.0` is compared with an absolute strict difference of one epsilon).

Also, the sin-method is not slower than the cos-method on the [0.0, 1.0] interval, at least with a naive scalar implementation.

So, the new version passes the unit tests, and should essentially extract most of the available precision, without sacrificing much performance.

**Comment by qsantos** [comment]
To visualize better what this PR does, this is the imprecision of [`acos(cos(x))`](https://herbie.uwplse.org/demo/330a905ec0693db542db68460d3bd20e2a5cc455.f316bcef3de0492d34dcdbc4c663eb04d00305c4/graph.html):


We easily notice the loss of precision in [-1; 1]. The change in behavior looks sharp, but to note that the x-axis is given in logarithmic scale. For |x| > 1, ULP(x) ≥ ε; since |cos(x)| ≤ 1, it does not waste precision. However, for |x| < 1, ULP(x) < ε (and particularily for |x| ≪ 1, ULP(x) ≪ ε); in this case ½ ≤ |cos(x)| < 1, so ULP(cos(x)) = ε / 2. It means that, as x gets close to 0, the absolute precision of its float32 representation increases, while that of cos(x) remains constant. In layman's terms, more and more possible values for x gets mapped to the same value by `cos`.

Outside the ±1 boundary, you get most of the precision back, so the curve looks almost flat. Within it, you quickly gets >100 % error, so it plateaus at what looks like 62 bits of error (maybe 64-bit fixed-point representation, 1 bit of sign, and 1 bit because a broken clock is right twice a day?). Between roughly ±1e-19 all precision is lost (even with 64 bits of precision), and acos(cos(x)) always return exactly 0. However, the relative precision gets lower, as the real values gets closer and closer to the approximated 0.

Compare this to the curve for [`asin(sin(x))`](https://herbie.uwplse.org/demo/6ee8603a30eb33159c1ae32f5e9e53062d3bedbb.f316bcef3de0492d34dcdbc4c663eb04d00305c4/graph.html):


Because `sin(x) ~ x` around 0, we observe no large loss of precision around there. Outside ±1, we have the same behavior as `acos(cos(x))`.

---

Although `asin(sin(x))` behaves much more nicely than `acos(cos(x))`, the sin-method of retrieving the angle of a quaternion also implies computing √(x²+y²+z²). The loss of precision of the full [sin-method](https://herbie.uwplse.org/demo/3ccaa652b56401fa4a7adc156f23af11cfb2b99f.f316bcef3de0492d34dcdbc4c663eb04d00305c4/graph.html) is shown below (note that it actually depends on the axis used):


Clearly, we are still losing quite a few bits of precision. So it's best to use the cos-method outside ±1, and the sin-method inside ±1.

Using `hypot` can improve things by replacing `sqrt(x*x + y*y + z*z)` with [`hypot(hypot(x, y), z)`](https://herbie.uwplse.org/demo/0e90e0d40a6ff2a0aab3b8d9665150d7cf1a83cb.f316bcef3de0492d34dcdbc4c663eb04d00305c4/graph.html). The result looks almost too good:


By manually iterating over the floats, it seems like the sin-hypot method indeed behaves much more nicely than the sin-sqrt method for values lower than roughly 2⁻⁶². Actually, it behaves well even on subnormal numbers, and only starts to break down below 2⁻¹²⁶. My guess is that this might be related to the 64 bits of the extended precision of the FPU. However, the asin-hypot method is several times slower than the asin-sqrt method where it matters. It might make sense to use the asin-hypot method below a set threshold, but the appropriate value for this threshold would most likely depend on the architecture.

File: glm/ext/quaternion_exponential.inl [hunk 1]
```diff
@@ -46,17 +48,34 @@ namespace glm
...
 
-		return qua<T, Q>(cos(NewAngle) * magnitude * Mag, x.x * Div * Mag, x.y * Div * Mag, x.z * Div * Mag);
+			T SinAngle = sqrt(VectorMagnitude) / magnitude;
+			T Angle = asin(SinAngle);
+			T NewAngle = Angle * y;
+			T Div = sin(NewAngle) / SinAngle;
+			T Mag = pow(magnitude, y - static_cast<T>(1));
+			return qua<T, Q>(cos(NewAngle) * magnitude * Mag, x.x * Div * Mag, x.y * Div * Mag, x.z * Div * Mag);
+		}
+		else
```
Comment (domgho) [inline glm/ext/quaternion_exponential.inl]: Redundant else due to return. Same below

Comment (qsantos) [reply glm/ext/quaternion_exponential.inl]: I feared that factoring the code would make it harder to read here. But you are right that it does duplicate quite a few formulas. I have pushed a new version that takes your remark into account.

File: glm/ext/quaternion_trigonometric.inl [hunk 2]
```diff
@@ -1,9 +1,18 @@
...
 {
 	template<typename T, qualifier Q>
 	GLM_FUNC_QUALIFIER T angle(qua<T, Q> const& x)
 	{
-		return acos(x.w) * static_cast<T>(2);
+		if (abs(x.w) > cos_one_over_two<T>())
+		{
+			return asin(sqrt(x.x * x.x + x.y * x.y + x.z * x.z)) * static_cast<T>(2);
+		}
+		else
```
Comment (domgho) [inline glm/ext/quaternion_trigonometric.inl]: Nice, this else can be still removed though due to the return in the if branch

Comment (qsantos) [reply glm/ext/quaternion_trigonometric.inl]: Done.

File: glm/ext/quaternion_exponential.inl (hunk 1 above)
Comment (domgho) [reply glm/ext/quaternion_exponential.inl]: Are you sure this will compile like this? Angle is not defined below

Comment (qsantos) [reply glm/ext/quaternion_exponential.inl]: Of course! My bad, I should have run the tests before pushing. Fixed.

Comment (domgho) [reply glm/ext/quaternion_exponential.inl]: Did you push the fix?

Comment (qsantos) [reply glm/ext/quaternion_exponential.inl]: Apparently I didn't :sweat: . Sorry for that.

**Comment by qsantos** [comment]
Finally placated the gods of the CI :sweat_smile:

**Comment by Groovounet** [comment]
I really wish all the PR would be of that level of quality !

Thanks for contributing !
Christophe

**Comment by qsantos** [comment]
Thank you very much for your patience and your kinds words. GLM is really a great library, I was happy to be able to contribute!
